### PR TITLE
Fix input for (L)BFGS and PSO

### PR DESCRIPTION
**Pull request by pkofod**
ref #399

### CODE REVIEW

File: src/multivariate/solvers/first_order/bfgs.jl [hunk 1]
```diff
@@ -72,11 +72,10 @@ function update_state!{T}(d, state::BFGSState{T}, method::BFGS)
     copy!(state.x_previous, state.x)
 
     # Update current position
-    @simd for i in 1:n
-        @inbounds state.dx[i] = state.alpha * state.s[i]
-        @inbounds state.x[i] = state.x[i] + state.dx[i]
+    @simd for i = 1:n
+        state.dx[i] .= state.alpha .* state.s[i]
+        state.x[i] .= state.x[i] .+ state.dx[i]
```
Comment (anriseth) [inline src/multivariate/solvers/first_order/bfgs.jl]: Will it not work to do `state.x .= state.x .+ state.dx`? (Would dot-notation not work for arbitrary vector-types?)

Comment (anriseth) [inline src/multivariate/solvers/first_order/bfgs.jl]: Sorry, I got confused. I mean on the `state.s`: I thought that's meant to be a vector, whilst `state.x` may be a multidimensional array?

Comment (pkofod) [inline src/multivariate/solvers/first_order/bfgs.jl]: This is a mistake! I had originally tried to change it to broadcast, but that didn't like mixing vectors and matrices. 

I mean, `state.x` is allowed to be multidimensional. If `x` is passed as a matrix we want `Optim.minimizer(res)` to be a matrix as well.

We could store `s` as the same type as `x`, and throw `vec`'s around `state.s` whereever it shows up?

**Comment by anriseth** [comment]
Would reshape on state.s work?

File: src/multivariate/solvers/first_order/bfgs.jl (hunk 1 above)
Comment (anriseth) [inline src/multivariate/solvers/first_order/bfgs.jl]: Would reshape on state.s work?
Alternatively reshape on `Optim.minimizer(res)`, and then just have `state.x` as a vector internally

**Comment by pkofod** [comment]
> Would reshape on state.s work?

That's what my `vec` suggestion does, so yes! and I'll do just that

File: src/multivariate/solvers/first_order/bfgs.jl (hunk 1 above)
Comment (pkofod) [inline src/multivariate/solvers/first_order/bfgs.jl]: yeah, ofc you could just do `minimizer = similar(initial_x); copy!(minimizer, state.x)` in the end, that's another possibility.

File: test/multivariate/array.jl [hunk 2]
```diff
@@ -9,36 +9,37 @@
...
-                _minimizer = Optim.minimizer(res)
-                @test typeof(_minimizer) <: Array{Float64, 3}
-                @test size(_minimizer) == (2,2,1)
+        for m in (AcceleratedGradientDescent, BFGS, LBFGS, ConjugateGradient,  GradientDescent, MomentumGradientDescent, ParticleSwarm, SimulatedAnnealing)
+            res = optimize(f, g!, eye3, m())
+            _minimizer = Optim.minimizer(res)
+            @test typeof(_minimizer) <: Array{Float64, 3}
+            @test size(_minimizer) == (2,2,1)
+            if !(m in (SimulatedAnnealing, ParticleSwarm))
+                println(typeof(m))
```
Comment (pkofod) [inline test/multivariate/array.jl]: yes yes, this is debugging text, will remove... :)

**Pull request merged**